### ISSUE TITLE
/appRoleAssignments lacks paging or a way to retrieve more than 999 assignments if they exist

### ISSUE THREAD

**Issue by JoeKaplan** [microsoftgraph/microsoft-graph-docs#5877]
In my testing in my org's prod tenant that has a group with over 1400 assignments, the /groups/{id}/appRoleAssignments query does not seem to be able to return them all nor does it provide any notice that more data may be available. In the default query I get a page of 100 results. If I add $top=999 (the max page size allowed here), I can get the first 999. However, neither query provides a next page link or indicates that any data might be missing. This seems to be a missing feature in the API but it also leads to misleading results since the API doesn't tell you that there was more data available.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c3546a45-22f6-8f62-9f5d-26450c484572
* Version Independent ID: 0f6521fa-e1ed-ae22-9958-6a118c3518e6
* Content: [Get appRoleAssignment - Microsoft Graph beta](https://docs.microsoft.com/en-us/graph/api/approleassignment-get?view=graph-rest-beta&tabs=http#feedback)
* Content Source: [api-reference/beta/api/approleassignment-get.md](https://github.com/microsoftgraph/microsoft-graph-docs/blob/master/api-reference/beta/api/approleassignment-get.md)
* Product: ****
* Technology: **microsoft-graph**
* GitHub Login: @jthake-msft
* Microsoft Alias: **MSGraphDocsVteam**

**Comment by jthake-msft** [comment]
@davidmu1 do you know anyone that could help here?

**Comment by Gizmotist** [comment]
Ohh I could see this being an issue as well for the org I'm in, +1 on wanting to know a solution

**Comment by sfjverbarg** [comment]
We've hit that 100 limit... were wondering why we didn't get the next page link.

**Comment by Bhavya207** [comment]
@psignoret Could you please respond to @sfjverbarg  question.

**Comment by sfjverbarg** [comment]
@psignoret In discussions with @Bhavya207, We've found this bug doesn't just affect the Microsoft Graph beta API, but it also affects the Azure AD Graph API as well. Since we have no api that will work once we have over 999 roles assigned to a user, this is expected to become a problem within the next couple of months for us. We need to get a commitment on when this will be resolved.

**Comment by Bhavya207** [comment]
> @psignoret In discussions with @Bhavya207, We've found this bug doesn't just affect the Microsoft Graph beta API, but it also affects the Azure AD Graph API as well. Since we have no api that will work once we have over 999 roles assigned to a user, this is expected to become a problem within the next couple of months for us. We need to get a commitment on when this will be resolved.

+++++++++++++++++++++
@psignoret Can you please respond here

**Comment by psignoret** [comment]
@sfjverbarg We do not have a committed time at time point for when full paging for both direct and indirect app role assignments will be available. In your scenario, are all user assignments direct (user is assigned directly), or are they also indirect (i.e. user is a member of group, group is assigned directly).

**Comment by sfjverbarg** [comment]
@psignoret 99% of user assignments are indirect via group membership.

**Comment by psignoret** [comment]
@sfjverbarg Thanks for the follow up. 

Providing full paging support for all indirect app role assignments is going to take some time. In the interim, we're considering the following stop-gap approach, and I'd like to see what you think of this:

(NOTE: The filters shown in the following examples are illustrative. They don't actually work today, and they may look different in the future.)

1. Add full paging support for _direct_ app role assignments on a user:
   
   ```http
   GET /v1.0/users/{user-id}/appRoleAssignments?$filter=principalId eq {user-id}
   ```

2. Add support for identifying all groups that a user is a direct member of which have at least one app role assignment:
   
   ```http
   GET /v1.0/users/{user-id}/memberOf?$filter=appRoleAssignments/$count gt 0
   ```

3. Add full paging support for app role assignments on a group (which are always direct):

   ```http
   GET /v1.0/groups/{group-id}/appRoleAssignments
   ```

With these three, you would be able to expand the full set of direct and indirect app role assignments for a user as follows:

1. Get all direct assignments for the user.
2. Get all groups the user is a direct member of which have at least all app role assignment.
3. For each of these groups, get all direct app role assignments.

Supposing you had some high-quality sample code and scripts for how to do this optimally using [batch requests](https://docs.microsoft.com/graph/json-batching), would an approach like this be something that would work for you?

**Comment by sfjverbarg** [comment]
This looks like it would work. Is it possible to get a time frame on when this would be available?

**Comment by psignoret** [comment]
@sfjverbarg Thanks for the feedback. We don't have a timeline to share yet, but we're actively working on the first point now.

**Comment by Yash27101998** [comment]
Is this issue still exist ?

**Comment by psignoret** [comment]
@Yash27101998  Yes. We're making progress on the plan described above, but we still don't have a specific date to share. Can you share details about what you're trying to do?

**Comment by wtfstoops** [comment]
what is the status of this request?  We are company of well over 90,000 people.  We have long exceeded the limit for 100 returns only.

**Comment by ryanspletzer** [comment]
So IIRC last year there was an attempt to role this out but there was a bug and it was pulled back. Has there been any attempt to re-deploy this?

**Comment by davidmu1** [comment]
No change identified for the page. Closing the issue.

**Comment by accenture-eso** [comment]
This result is confusing. It does not appear that the API has changed such that /user/appRoleAssignments supports paging as requested but the issue was closed. Does this mean that Microsoft will is rejecting the request to add this feature or that Microsoft believes the issue was fixed by adding this feature? Or is this tracked somewhere else? Thanks!

**Comment by psignoret** [comment]
Reopening until we have something else to point people to, as this has caused some confusion.

@accenture-eso (and everyone else who is wondering): The changes that will allow the query pattern [I described earlier](https://github.com/microsoftgraph/microsoft-graph-docs/issues/5877#issuecomment-637449037) has been implemented and tested out internally, and is on it's way to beta. (Keep in mind, though, that "on it's way" could still mean several weeks, due to our SDP (~slow~ safe deployment process).)

**Comment by FaithOmbongi** [comment]
Hello all,

**UPDATE:** Paging is now supported for /appRoleAssignments as indicated in this changelog release: [added support for paging appRoleAssignments](https://developer.microsoft.com/en-us/graph/changelog/?search=paging#479558a4-5491-4463-bf2f-4474590767dbbeta). cc @psignoret 

Closing this issue.

**Comment by psignoret** [comment]
Reopening, as there are still some pending documentation updates needed to explain the limitations of the new paging capabilities and additional details on how to get the full list.

**Comment by lramosvea** [comment]
Hi @psignoret @FaithOmbongi, what's the status of this issue? Have docs been updated already? If so, can we close this issue?

**Comment by lramosvea** [comment]
@FaithOmbongi could you help with this doc? Not sure if it no longer exists. Thanks.

**Comment by lramosvea** [comment]
@FaithOmbongi should we close this issue? I can't seem to find the doc.

**Comment by FaithOmbongi** [comment]
> Reopening, as there are still some pending documentation updates needed to explain the limitations of the new paging capabilities and additional details on how to get the full list.

I'm closing this issue on GitHub and tracking the pending doc updates through internal work item number 17552 in the onboarding ADO assigned to @psignoret. Thanks @lramosvea .
